lib/Utils/PlatformWin.cpp

```diff
@@ -0,0 +1,687 @@
+
+    const char *keyEnd = placeHolder - 1;
+    const char *nextKey = placeHolder;
+    // Find end of previous key.
+    while ((keyEnd > subKey) && (*keyEnd != '\\'))
+      keyEnd--;
+    // Find end of key containing $VERSION.
+    while (*nextKey && (*nextKey != '\\'))
+      nextKey++;
+    size_t partialKeyLength = keyEnd - subKey;
+    char partialKey[256];
+    if (partialKeyLength > sizeof(partialKey))
+      partialKeyLength = sizeof(partialKey);
+    ::strncpy(partialKey, subKey, partialKeyLength);
+    partialKey[partialKeyLength] = '\0';
+    HKEY hTopKey = NULL;
+    lResult = ::RegOpenKeyExA(hRootKey, partialKey, 0,
+                              KEY_READ | KEY_WOW64_32KEY, &hTopKey);
+    if (lResult == ERROR_SUCCESS) {
+      char keyName[256];
+      // int bestIndex = -1;
+      double bestValue = 0.0;
+      DWORD size = sizeof(keyName) - 1;
+      for (DWORD index = 0; ::RegEnumKeyExA(hTopKey, index, keyName, &size, NULL,
+                                            NULL, NULL, NULL) == ERROR_SUCCESS;
+           index++) {
+        const char *sp = keyName;
+        while (*sp && !isdigit(*sp))
+          sp++;
+        if (!*sp)
+          continue;
+        const char *ep = sp + 1;
+        while (*ep && (isdigit(*ep) || (*ep == '.')))
+          ep++;
+        char numBuf[32];
+        ::strncpy(numBuf, sp, sizeof(numBuf) - 1);
+        numBuf[sizeof(numBuf) - 1] = '\0';
+        double dvalue = ::strtod(numBuf, NULL);
+        if (dvalue > bestValue) {
+          // Test that InstallDir is indeed there before keeping this index.
+          // Open the chosen key path remainder.
+          ::strcpy(bestName, keyName);
+          // Append rest of key.
+          ::strncat(bestName, nextKey, sizeof(bestName) - 1);
+          bestName[sizeof(bestName) - 1] = '\0';
+          lResult = ::RegOpenKeyExA(hTopKey, bestName, 0,
+                                    KEY_READ | KEY_WOW64_32KEY, &hKey);
+          if (lResult == ERROR_SUCCESS) {
+            if (readFullStringValue(hKey, valueName, outValue)) {
+              // bestIndex = (int)index;
+              bestValue = dvalue;
+              returnValue = true;
+            }
+            ::RegCloseKey(hKey);
+          }
+        }
+        size = sizeof(keyName) - 1;
+      }
+      ::RegCloseKey(hTopKey);
+    } else
+      ReportError(lResult, "RegOpenKeyEx");
+  } else {
+    // If subKey is empty, then valueName is subkey, and we retreive that
+    if (subKey[0]==0) {
+      subKey = valueName;
+      valueName = nullptr;
+    }
+    lResult = ::RegOpenKeyExA(hRootKey, subKey, 0, KEY_READ | KEY_WOW64_32KEY,
+                              &hKey);
+    if (lResult == ERROR_SUCCESS) {
+      returnValue = readFullStringValue(hKey, valueName, outValue);
+      ::RegCloseKey(hKey);
+    } else
+      ReportError(lResult, "RegOpenKeyEx");
+  }
+  return returnValue;
+}
+
+int GetVisualStudioVersionCompiledWith() {
+#if (_MSC_VER >= 1900)
+  return 14;
```

Comment:
The MSDN suite claims:

> Each major release of the Visual C++ compiler increments the “hundreds” of _MSC_VER. Each update within a major release increments the “units” by 1. For example, in VS “15” Preview 5, the macro _MSC_VER evaluates to 1910. The next update will have _MSC_VER set to 1911.
> 
> Note that VS “15” and VS2015 are both major releases of Visual Studio with different major version numbers. The included compiler toolset, however, will have the same major version number – with the change described here, the minor version number can be used to distinguish compiler toolsets.

From which I am concluding that we should *not* care about the VisualStudio number at all but only about the 'compiler toolset' version number ...  In part because the link between the two version is going to be completely arbitrary ...

But then again, I have to ask :) how are we using this function GetVisualStudioVersionCompiledWith?